src/generic/fetch.ts

```diff
@@ -0,0 +1,57 @@
+/**
+ * Contains utility methods using 'fetch'.
+ */
+
+import {ApiError, ErrorCode} from '../core/api_error';
+import {BFSCallback} from '../core/file_system';
+
+/**
+ * Asynchronously download a file as a buffer or a JSON object.
+ * Note that the third function signature with a non-specialized type is
+ * invalid, but TypeScript requires it when you specialize string arguments to
+ * constants.
+ * @hidden
+ */
+export function fetchFileAsync(p: string, type: 'buffer', cb: BFSCallback<Buffer>): void;
+export function fetchFileAsync(p: string, type: 'json', cb: BFSCallback<any>): void;
+export function fetchFileAsync(p: string, type: string, cb: BFSCallback<any>): void;
+export function fetchFileAsync(p: string, type: string, cb: BFSCallback<any>): void {
+  fetch(p)
+    .then((res) => {
+      if (!res.ok) {
+        return cb(new ApiError(res.status, "fetch error."));
+      } else {
+        switch (type) {
+          case 'buffer':
+            res.arrayBuffer()
+              .then((buf) => cb(null, Buffer.from(buf)))
+              .catch((err) => cb(new ApiError(0, err.message)));
+            break;
+          case 'json':
+            res.json()
+              .then((json) => cb(null, json))
+              .catch((err) => cb(new ApiError(0, err.message)));
+            break;
+          default:
+            throw new ApiError(ErrorCode.EINVAL, "Invalid download type: " + type);
+        }
+      }
+    })
+    .catch((err) => cb(new ApiError(0, err.message)));
```

Comment:
Oh, sorry for the ambiguity. I meant `.catch`'ed -- meaning, a rejected promise. (We shouldn't `try / catch` this; that seems like a truly exceptional error.)

I could not find anything in the spec (in my brief search) concerning when the promise gets rejected, and if it gets rejected with an error object.

Regardless, it seems like it should always be an `EIO` error anyway, as all of the other relevant checks occur on the file index.